### PR TITLE
JSX Docs: JSX In Depth and React Without JSX

### PR DESCRIPTION
**Pull request by lacker**
This creates two new documents. "JSX In Depth" attempts to describe all of JSX, covering the occasionally-relevant edge cases and weird bits. It should be a document you can go to, to get your weird JSX questions answered. "React Without JSX" just explains how to use React without JSX.

Some docs in the old site are now obsoleted. "JSX Spread Attributes" and "JSX Gotchas" along with a number of the JSX-specific tips are just part of "JSX In Depth" now.

I just removed the translations of the no-longer-there docs. Dunno what else to do with them.

Thanks @spicyj for answering a zillion of my JSX questions in person this afternoon ;-) But still I'm sure I am mis-explaining some things here, so I'd appreciate a sharp-eyed review of this content.

### CODE REVIEW

**Comment by lacker** [comment]
screencapture-localhost-4000-react-docs-react-without-jsx-html-1475641564288 1

**Comment by lacker** [comment]
screencapture-localhost-4000-react-docs-jsx-in-depth-html-1475641546630 1

File: docs/docs/jsx-in-depth.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+You can also use the self-closing form of the tag if there are no children. So:
+
+```js
+<div id={'blorp'} />
+```
+
+transpiles into:
+
+```js
+React.createElement('div', {id: 'blorp'}, null)
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Minor not: you'd almost never want to use the `id` attribute in React components. Maybe `className`?

File: docs/docs/jsx-in-depth.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+If you want to test out how some specific JSX is converted into JavaScript, you can try out [the online Babel transpiler](http://tinyurl.com/j26hzbz).
+
+## HTML Tags and React Components
+
+The first part of a JSX tag determines the type of the JSX element.
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: "Of the React element"? We don't really say "JSX element" elsewhere.

File: docs/docs/jsx-in-depth.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+```js
+import React from 'react';
+import MyComponent from './MyComponent';
+
+export default function() {
+  return <MyComponent />;
+}
+```
+
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Maybe add here:

"If you don't use a JavaScript bundler and added React as a script tag, it is already in scope as a `React` global."

File: docs/docs/jsx-in-depth.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+You can also refer to a React component using dot-notation from within JSX. This is convenient if you have a single module that exports many React components. For example, if you have a `Settings` module that exports a `Settings.DatePicker` component, you can invoke it directly from JSX with:
+
+```js
+import React from 'react';
+import Settings from './settings';
+
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: I feel like this is a minefield because it only works if you use default exports in that file. Maybe shift the attention from imports to demonstrating this in one file? So that we don't have to explain how modules work in this document.

File: docs/docs/jsx-in-depth.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+export default function() {
+  return <Settings.DatePicker barColor={'blue'} />;
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Let's give names to all functions.

File: docs/docs/jsx-in-depth.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+import Settings from './settings';
+
+var DatePicker = Settings.DatePicker;
+
+export default function() {
+  return <DatePicker barColor={'blue'} />;
+}
+```
+
+If the element type starts with a lowercase letter, React will create a HTML element. In general, name your components starting with capital letters, and this will be fine. If you do have a component that starts with a lowercase letter, assign it to a capitalized variable before using it in JSX.
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: This sounds a bit like React will create an actual HTML node when you type JSX. It is still a "React element" but it describes a DOM tag.

File: docs/docs/jsx-in-depth.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+import React from 'react';
+
+function hello(props) {
+  // This use of <div> is legitimate because div is a valid HTML tag
+  return <div>Hello {props.toWhat}</div>;
+}
+
+export default function() {
+  // This code attempts to create an HTML <hello> tag and fails
+  return <hello {toWhat='World'} />
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Unintentional braces

File: docs/docs/jsx-in-depth.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+```js
+import React from 'react';
+import findComponent from './findComponent';
+
+function render1() {
+  // Not valid JSX
+  return <findComponent() />;
+}
+
+var MyComponent = findComponent();
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Let's move this inside the function? That would be equivalent to your first example then.

File: docs/docs/jsx-in-depth.md [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+```js
+<MyComponent message="hello world" />
+
+<MyComponent message={"hello world"} />
+```
+
+When you pass a string literal, its value is HTML-unescaped. So these two JSX expressions are equivalent:
+
+```js
+<MyComponent message="&lt;3" />
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Today I learned

File: docs/docs/jsx-in-depth.md [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+  <ul>
+    <li>Item 1</li>
+    <li>Item 2</li>
+  </ul>
+</div>
+```
+
+A React component can't return multiple JSX elements, but a single JSX element can have multiple children, so if you want a component to render multiple things you can wrap it in a `div` like this.
+
+When you pass JSX as the children, `props.children` does not actually let you access the child component from the parent. Instead, it lets you access the return value of `React.createElement` calls using the `React.Children` helper library. Generally, accessing children in this way is only useful to return in a render function, rather than modifying or altering the children.
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: I don't understand this paragraph, can you rephrase (maybe with an example)

File: docs/docs/jsx-in-depth.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+`false` or `null` are valid children. They simply don't render. These JSX expressions will all render to the same thing:
+
+```js
+<div />
+
+<div></div>
+
+<div>{false}</div>
+
+<div>{null}</div>
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: This document should also include detailed information on how whitespace woks in JSX.

File: docs/docs/jsx-in-depth.md (hunk 4 above)
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: In general, "dot notation" for components is not popular and I think could live somewhere in the bottom of the document. We don't need to emphasize this, and other topics (such as how expressions and whitespace work) seem more valuable to me.

File: docs/docs/jsx-in-depth.md (hunk 1 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

**Comment by spicyj** [comment]
No tinyurl please -- no need to add another dependency.

File: docs/docs/jsx-in-depth.md (hunk 4 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: I ditched the import and cut it down to one example instead of two to de-emphasize it. I'd like to keep this here though because this section of the document is explaining "all the ways to specify the element type".

File: docs/docs/jsx-in-depth.md (hunk 6 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK rewrote this part

File: docs/docs/jsx-in-depth.md (hunk 7 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: Fixed

File: docs/docs/jsx-in-depth.md (hunk 8 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md (hunk 10 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: I'm just going to delete it - it probably belongs more appropriately in somewhere that discusses "components vs instances vs elements" rather than "all about jsx" because it's not really jsx-specific

File: docs/docs/jsx-in-depth.md [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+```js
+function renderItem(message) {
+  return <li>{message}</li>;
+}
+
+function renderTodoList() {
+  var todos = ['finish doc', 'submit pr', 'nag dan to review'];
+  return (
+    <ul>
+      {todos.map(renderItem)}
```
Comment (keyanzhang) [inline docs/docs/jsx-in-depth.md]: Since this is an array, each child should have a unique `key` prop.

Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

File: docs/docs/jsx-in-depth.md (hunk 11 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK I added a section about whitespace

File: docs/docs/jsx-in-depth.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+function Repeat(props) {
+  // The numbers 0 .. (numTimes - 1)
+  var range = [...Array(props.numTimes).keys()];
+  return <div>{range.map(props.children)}</div>
+}
+
+function ListOfTenThings() {
+  return (
+    <Repeat>
+      {(index) => <div>This is item {index} in the list</div>}
```
Comment (keyanzhang) [inline docs/docs/jsx-in-depth.md]: `<Repeat numTimes={10}>`? :) Also missing `key` here.

File: docs/docs/jsx-in-depth.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+`if` statements and `for` loops are not expressions in JavaScript, so they can't be used in JSX directly. Instead, you can put these in the surrounding code. For example:
+
+```js
+function NumberDescriber(props) {
+  var description;
+  if (props.number % 2 == 0) {
+    word = 'even';
+  } else {
+    word = 'odd';
```
Comment (keyanzhang) [inline docs/docs/jsx-in-depth.md]: did you mean `description = 'odd'`?

File: docs/docs/jsx-in-depth.md (hunk 13 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: Fixed

File: docs/docs/jsx-in-depth.md (hunk 14 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

**Comment by lacker** [comment]
OK, I responded to everything here.

File: docs/docs/jsx-in-depth.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
+---
+id: jsx-in-depth
+title: JSX In Depth
+permalink: docs/jsx-in-depth.html
+---
+
+Fundamentally, JSX just provides syntactic sugar for the `React.createElement(component, props, children)` function. The JSX code:
+
+```js
+<MyComponent prop1={'one'} prop2={'two'}>stuff</MyComponent>
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: It's a little weird to me that this uses {} since people don't ever do that in practice.

File: docs/docs/jsx-in-depth.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+    return <div>imagine a {props.color} datepicker here</div>;
+  }
+}
+
+export default function BlueDatePicker() {
+  return <MyComponents.DatePicker color={'blue'} />;
+}
+```
+
+If the element type starts with a lowercase letter, React will try to create a React element that maps directly to an HTML tag, rather than creating an element based on a custom component. In general, name your components starting with capital letters, and this will be fine. If you do have a component that starts with a lowercase letter, assign it to a capitalized variable before using it in JSX.
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: Can we do something closer to:

> When an element type starts with a lowercase letter, it refers to a built-in component like `<div>` or `<span>` and results in a string `'div'` or `'span'` passed to React.createElement. Types that start with a capital letter like `<Foo />` compile to `React.createElement(Foo)` and correspond to a component defined or imported in your JavaScript file.
> 
> We recommend naming components with a capital letter. If you do have a component that starts with a lowercase letter, assign it to a capitalized variable before using it in JSX.

I don't want to over-pivot on HTML/DOM in the JSX docs.

File: docs/docs/jsx-in-depth.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+  return <div>Hello {props.toWhat}</div>;
+}
+
+export default function HelloWorld() {
+  // This code attempts to create an HTML <hello> tag and fails
+  return <hello toWhat="World" />
+}
+```
+
+You cannot use a general expression as the React element type. If you do want to use a general expression to indicate the type of the element, just assign it to a capitalized variable first. For example, if you have a `findComponent()` function that returns a component, and you want to render one:
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: A more realistic example here would be

```
const storyComponents = {
  photo: <PhotoStory />,
  video: <VideoStory />,
};

<storyComponents[this.props.storyType] />
```

This is where I most frequently see people actually having problems.

File: docs/docs/jsx-in-depth.md [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+
+For `MyComponent`, The value of `props.foo` will be `10` because the expression `1 + 2 + 3 + 4` gets evaluated.
+
+`if` statements and `for` loops are not expressions in JavaScript, so they can't be used in JSX directly. Instead, you can put these in the surrounding code. For example:
+
+```js
+function NumberDescriber(props) {
+  var description;
+  if (props.number % 2 == 0) {
+    description = 'even';
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: Can we use an example where the branches each make a component? Maybe just `<strong>even</strong>`.

File: docs/docs/jsx-in-depth.md [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+
+If you pass no value for a prop, it defaults to `true`. These two JSX expressions are equivalent:
+
+```js
+<MyTextBox autocomplete />
+
+<MyTextBox autocomplete={true} />
+```
+
+In general, we don't recommend using this. It's more consistent to just use the second form and explicitly pass `true`. This behavior is just there so that it matches the behavior of HTML.
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: In general, we don't recommend using this form because it can be confused with the ES6 object shorthand `{autocomplete}` which is short for `{autocomplete: autocomplete}` rather than `{autocomplete: true}`.

File: docs/docs/jsx-in-depth.md [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+<MyComponent>Hello world!</MyComponent>
+```
+
+This is valid JSX, and `props.children` in `MyComponent` will simply be the string `"Hello world!"`. HTML is unescaped, so you can generally write JSX just like you would write HTML in this way:
+
+```html
+<div>This is valid HTML &amp; JSX at the same time.</div>
+```
+
+JSX removes whitespace at the beginning and ending of a line. It also removes blank lines. So these all render to the same thing:
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: Not quite true when two pieces of text adjoin:

```
<div>
Hello
World
</div>
```

gives "Hello World".

File: docs/docs/jsx-in-depth.md [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+<div>
+
+  Hello
+
+</div>
+```
+
+### JSX Children
+
+You can provide more JSX as the children. This is useful for displaying nested components:
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: more JSX elements as

File: docs/docs/jsx-in-depth.md [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+
+```js
+// Repeats a component numTimes
+function Repeat(props) {
+  // The numbers 0 .. (numTimes - 1)
+  var range = [...Array(props.numTimes).keys()];
+  return <div>{range.map(props.children)}</div>
+}
+
+function ListOfTenThings() {
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: Can we swap these two functions? Easier to understand with ListOfTenThings I think.

File: docs/docs/jsx-in-depth.md [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+}
+```
+
+Normally, JavaScript expressions inserted in JSX will evaluate to a string, a React element, or a list of those things. However, `props.children` works just like any other prop in that it can pass any sort of data, not just the sorts that React knows how to render. For example, if you have a custom component, you could have it take a callback as `props.children`:
+
+```js
+// Repeats a component numTimes
+function Repeat(props) {
+  // The numbers 0 .. (numTimes - 1)
+  var range = [...Array(props.numTimes).keys()];
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: This is hard to understand. Why not a for loop?

File: docs/docs/jsx-in-depth.md [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+    <Repeat numTimes={10}>
+      {(index) => <div key={index}>This is item {index} in the list</div>}
+    </Repeat>
+  );
+}
+```
+
+This usage is not common, but it works if you want to stretch what JSX is capable of.
+
+`false` or `null` are valid children. They simply don't render. These JSX expressions will all render to the same thing:
```
Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: It would be helpful to explain why this is useful `<div>{showHeader && <Header />}<Content /></div>`.

P.S. true also is ignored. This is not JSX but is React DOM and might not belong here.

File: docs/docs/react-without-jsx.md [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+ReactDOM.render(<Hello toWhat={'World'} />,
+                document.getElementById('root'));
+```
+
+can also be written without JSX:
+
+```js
+class Hello extends React.Component {
```
Comment (spicyj) [inline docs/docs/react-without-jsx.md]: I know you show it right below, but can we do `const e =` in every example except when we're specifically discussing what JSX compiles to? It's the recommended way to use createElement.

File: docs/docs/react-without-jsx.md [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+For example, this code written with JSX:
+
+```js
+class Hello extends React.Component {
+  render() {
+    return <div>Hello {this.props.toWhat}</div>;
+  }
+}
+
+ReactDOM.render(<Hello toWhat={'World'} />,
```
Comment (spicyj) [inline docs/docs/react-without-jsx.md]: fb style, please

```
ReactDOM.render(
  <Hello ... />,
  document...
);
```

File: docs/docs/react-without-jsx.md [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+can also be written without JSX:
+
+```js
+class Hello extends React.Component {
+  render() {
+    return React.createElement('div', null, `Hello ${this.props.toWhat}`);
+  }
+}
+
+ReactDOM.render(React.createElement(Hello, {toWhat: 'World'}, null),
```
Comment (spicyj) [inline docs/docs/react-without-jsx.md]: ditto

File: docs/downloads/single-file-example.html [hunk 28]
```diff
@@ -2,7 +2,7 @@
 <html>
   <head>
     <meta charset="UTF-8" />
-    <title>Hello React World!</title>
+    <title>hello</title>
```
Comment (spicyj) [inline docs/downloads/single-file-example.html]: intentional?

File: docs/docs/jsx-in-depth.md (hunk 2 above)
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Now that I read @spicyj comments I think it's fine (haha sorry). Since this is a doc about JSX specifically I think it's fine to say JSX Element.

File: docs/docs/jsx-in-depth.md [hunk 29]
```diff
@@ -0,0 +1,312 @@
+---
+id: jsx-in-depth
+title: JSX In Depth
+permalink: docs/jsx-in-depth.html
+---
+
+Fundamentally, JSX just provides syntactic sugar for the `React.createElement(component, props, children)` function. The JSX code:
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Can make this `...children`? It accepts variadic arguments.

File: docs/docs/jsx-in-depth.md (hunk 15 above)
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: That, and let's use double quotes in JSX too since most popular styleguides do that. So it'll be `prop1="one"`.

I would also like us to use more real-world over contrived examples. Like Button or Avatar instead of MyComponent or Foo.

File: docs/docs/jsx-in-depth.md [hunk 30]
```diff
@@ -0,0 +1,312 @@
...
+
+For example, both of the imports are necessary in this code, even though 'React' and 'MyComponent' are not directly referenced from JavaScript:
+
+```js
+import React from 'react';
+import MyComponent from './MyComponent';
+
+export default function MyRedComponent() {
+  return <MyComponent color="red" />;
+}
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Same here: could be Button and DangerButton.

File: docs/docs/jsx-in-depth.md [hunk 31]
```diff
@@ -0,0 +1,312 @@
...
+```
+
+If you don't use a JavaScript bundler and added React as a script tag, it is already in scope as a React global.
+
+You can also refer to a React component using dot-notation from within JSX. This is convenient if you have a single module that exports many React components. For example, if `MyComponents.DatePicker` is a component, you can invoke it directly from JSX with:
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Nit: "use" rather than "invoke" because you won't actually be calling it

File: docs/docs/jsx-in-depth.md [hunk 32]
```diff
@@ -0,0 +1,312 @@
...
+
+```js
+import React from 'react';
+
+function hello(props) {
+  // This use of <div> is legitimate because div is a valid HTML tag
+  return <div>Hello {props.toWhat}</div>;
+}
+
+export default function HelloWorld() {
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: I think we shouldn't emphasize exports in examples unless we specifically mention module boundaries. It's fine to leave this as a function and presume that it gets rendered by something later.

File: docs/docs/jsx-in-depth.md [hunk 33]
```diff
@@ -0,0 +1,312 @@
...
+
+function render2() {
+  var MyComponent = findComponent();
+
+  // Valid JSX
+  return <MyComponent />;
+}
+```
+
+## Props in JSX
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: I'm conflicted about calling them props here.

If we are talking about JSX, "attributes" seems more correct. Attributes usually become props (for custom components). However `key` and `ref` are attributes obeying JSX rules that don't become props. 

I don't feel strongly about this though. @spicyj What do you think?

File: docs/docs/jsx-in-depth.md (hunk 15 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

File: docs/docs/jsx-in-depth.md [hunk 34]
```diff
@@ -0,0 +1,312 @@
...
+
+When you pass a string literal, its value is HTML-unescaped. So these two JSX expressions are equivalent:
+
+```js
+<MyComponent message="&lt;3" />
+
+<MyComponent message={"<3"} />
+```
+
+This behavior is usually not relevant. It's useful for `children`, but not for most props. It's only mentioned here for completeness.
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Nit: it is not obvious to me how this is useful for children, or what children "content" prop has to do with explicitly specified string literal attributes.

File: docs/docs/jsx-in-depth.md (hunk 16 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK I just took this wording exactly

File: docs/docs/jsx-in-depth.md (hunk 12 above)
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: An inquisitive reader would think: when don't want to do this instead of using `<Item>`?

In general we want to encourage people to create components rather than call functions directly. But I guess this is fine.

I don't have a good solution to this.

Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Maybe if we put map body inline as an arrow it is clearer that this pattern is mostly useful for one-liners.

File: docs/docs/jsx-in-depth.md [hunk 35]
```diff
@@ -0,0 +1,314 @@
...
+
+function ListOfTenThings() {
+  return (
+    <Repeat numTimes={10}>
+      {(index) => <div key={index}>This is item {index} in the list</div>}
+    </Repeat>
+  );
+}
+```
+
```
Comment (gaearon) [inline docs/docs/jsx-in-depth.md]: Maybe add another post-explainer here: Children passed to a custom component can be anything as long as that component transforms them into something React can understand before rendering.

File: docs/docs/jsx-in-depth.md (hunk 17 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK I switched to use an example like this

File: docs/docs/jsx-in-depth.md (hunk 18 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: Fixed

File: docs/docs/jsx-in-depth.md (hunk 21 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md (hunk 22 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md (hunk 24 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: I like your example. Added

Hmm, this seems like a super logical place to mention that true is ignored... I'll just mention it here.

Comment (spicyj) [inline docs/docs/jsx-in-depth.md]: I meant all of true/false/null being ignored is a React DOM thing, not a JSX thing.

File: docs/downloads/single-file-example.html (hunk 28 above)
Comment (lacker) [inline docs/downloads/single-file-example.html]: OK

File: docs/docs/jsx-in-depth.md (hunk 29 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md [hunk 36]
```diff
@@ -0,0 +1,330 @@
...
+compiles into:
+
+```js
+React.createElement(MyComponent, {prop1: 'one', prop2: 2}, 'stuff')
+```
+
+You can also use the self-closing form of the tag if there are no children. So:
+
+```js
+<div className={'blorp'} />
```
Comment (vjeux) [inline docs/docs/jsx-in-depth.md]: nit: "blorp"

File: docs/docs/jsx-in-depth.md [hunk 37]
```diff
@@ -0,0 +1,330 @@
...
+import React from 'react';
+
+var MyComponents = {
+  DatePicker: function(props) {
+    return <div>imagine a {props.color} datepicker here</div>;
+  }
+}
+
+export default function BlueDatePicker() {
+  return <MyComponents.DatePicker color={'blue'} />;
```
Comment (vjeux) [inline docs/docs/jsx-in-depth.md]: ="blue"

File: docs/docs/jsx-in-depth.md (hunk 30 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md (hunk 32 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK****

File: docs/docs/jsx-in-depth.md (hunk 34 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: I tweaked the wording a bit. AFAICT this behavior is useful so that

<div>Foo &amp; Bar</div>

works the same way in HTML and in JSX - you need an unescape because JSX will re-escape it later. I'm guessing someone implemented this for children and just implemented it for all string literal props at the same time? Dunno though. Seems like a historical wart.

Comment (vjeux) [inline docs/docs/jsx-in-depth.md]: Yeah, i'm not sure what children has anything to do here :)

The two times I ran into are:
(1) trying to write a "

```js
<C prop="this is \"something\"" />
```

is a syntax error as the `\` is not escaping the `"`

(2) trying to use an unicode sequence


```js
<C prop="\u2192" />
```

will not display the unicode →, it's just going to display \u2192. Instead you need to do

```js
<C prop="&#8594;" />
// or
<C prop={"\u2192"} />
```

File: docs/docs/jsx-in-depth.md (hunk 35 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK - yeah I didn't intend that part to explain functions, only to explain loops, so I turned it into a one-liner. Also added that post-explainer

File: docs/docs/jsx-in-depth.md (hunk 31 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: OK

File: docs/docs/jsx-in-depth.md [hunk 38]
```diff
@@ -0,0 +1,330 @@
...
+A React component can't return multiple React elements, but a single JSX expression can have multiple children, so if you want a component to render multiple things you can wrap it in a `div` like this.
+
+### JavaScript Expressions
+
+You can pass any JavaScript expression as children, by enclosing it within `{}`. For example, these expressions are equivalent:
+
+```js
+<MyComponent>foo</MyComponent>
+
+<MyComponent>{"foo"}</MyComponent>
```
Comment (vjeux) [inline docs/docs/jsx-in-depth.md]: nit: we use single quote for js strings

File: docs/docs/jsx-in-depth.md (hunk 36 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: i just switched this to be more realistic rather than "blorp"

File: docs/docs/jsx-in-depth.md (hunk 37 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

File: docs/docs/jsx-in-depth.md (hunk 38 above)
Comment (lacker) [inline docs/docs/jsx-in-depth.md]: fixed

**Comment by vjeux** [comment]
🎉
